File: 0999-regions-cut-by-slashes/0999-regions-cut-by-slashes.py
class Solution(object):
    def __init__(self):
        self.parent = []
        self.sz = []

    def find_set(self, x):
        if self.parent[x] == x:
            return x
        self.parent[x] = self.find_set(self.parent[x])  # Path compression
        return self.parent[x]

    def make_union(self, x, y):
        a = self.find_set(x)
        b = self.find_set(y)
        if a == b:
            return True

        if self.sz[b] > self.sz[a]:
            a, b = b, a  # Swap to ensure the larger set is the parent

        self.parent[b] = a
        self.sz[a] += self.sz[b]
        return False

    def regionsBySlashes(self, grid):
        n = len(grid) + 1
        self.parent = [i for i in range(n * n)]
        self.sz = [1] * (n * n)

        for i in range(n):
            for j in range(n):
                if i == 0 or j == 0 or i == n - 1 or j == n - 1:
                    x = i * n + j
                    self.make_union(0, x)

        ans = 1
        for i in range(len(grid)):
            s = grid[i]
            for j in range(len(s)):
                c = s[j]
                if c == '/':
                    x = (i + 1) * n + j
                    y = i * n + (j + 1)
                    if self.make_union(x, y):
                        ans += 1
                elif c == '\\':
                    x = i * n + j
                    y = (i + 1) * n + (j + 1)
                    if self.make_union(x, y):
                        ans += 1

        return ans
        

File: 0999-regions-cut-by-slashes/0999-regions-cut-by-slashes.cpp

class Solution {
public:
    vector<int> parent, sz;

    int find_set(int x){
        if(parent[x] == x) return x;
        return parent[x] = find_set(parent[x]);
    }

    bool make_union(int x, int y){
        int a = find_set(x);
        int b = find_set(y);
        if(a == b) return true;
        if(sz[a] < sz[b]) swap(a, b);
        parent[b] = a;
        sz[a] += sz[b];
        return false;
    }

    int regionsBySlashes(vector<string>& grid) {
        int n = grid.size()+1;
        parent.resize(n*n);
        sz.resize(n*n);
        for(int i = 0; i<n*n; i++){
            parent[i] = i;
            sz[i] = 1;
        }

        vector<int> rank(n*n, -1);
        int ans = 1;
        for(int i = 0; i<n; i++){
            for(int j = 0; j<n; j++){
                if(i == 0 || j == 0 || i == n-1 || j == n-1){
                    int y = i*n + j;
                    make_union(0, y);
                }
            }
        }
        for(int i = 0; i<grid.size(); i++){
            string str = grid[i];
            for(int j = 0; j<str.size(); j++){
                if(str[j] == '/'){
                    int x = (i+1)*n + j;
                    int y = i*n + (j+1);
                    if (make_union(x, y))
                        ans++;
                }
                else if(str[j] == '\\'){
                    int x = (i + 1) * n + (j+1);
                    int y = i * n + (j);
                    if(make_union(x, y)) ans++;
                }   
            }
        }
        return ans;
    }
}; 
